ethexe/runtime/common/src/state.rs

```diff
@@ -785,52 +788,162 @@ impl DispatchStash {
+        Self {
+            value: ValueWithExpiry {
+                value: dispatch.value,
+                expiry,
+            },
+            payload: dispatch.payload.clone(),
+            origin: dispatch.origin,
+        }
+    }
+}
+
+#[derive(Clone, Default, Debug, Encode, Decode, PartialEq, Eq, derive_more::Into)]
+#[cfg_attr(feature = "std", derive(serde::Serialize, serde::Deserialize))]
+pub struct UserMailbox(BTreeMap<MessageId, MailboxMessage>);
+
+impl UserMailbox {
+    fn add(&mut self, message_id: MessageId, mailbox_message: MailboxMessage) {
+        let r = self.0.insert(message_id, mailbox_message);
+        debug_assert!(r.is_none())
+    }
+
+    fn remove(&mut self, message_id: MessageId) -> Option<MailboxMessage> {
+        self.0.remove(&message_id)
+    }
+
+    fn is_empty(&self) -> bool {
+        self.0.is_empty()
+    }
+
+    fn store<S: Storage>(self, storage: &S) -> MaybeHashOf<Self> {
+        MaybeHashOf((!self.0.is_empty()).then(|| storage.write_user_mailbox(self)))
+    }
+}
+
 #[derive(Clone, Default, Debug, Encode, Decode, PartialEq, Eq, derive_more::Into)]
 #[cfg_attr(feature = "std", derive(serde::Serialize, serde::Deserialize))]
 pub struct Mailbox {
-    inner: BTreeMap<ActorId, BTreeMap<MessageId, ValueWithExpiry<Value>>>,
+    inner: BTreeMap<ActorId, HashOf<UserMailbox>>,
     #[into(ignore)]
     #[codec(skip)]
     changed: bool,
 }
 
 impl Mailbox {
-    pub fn add(&mut self, user_id: ActorId, message_id: MessageId, value: Value, expiry: u32) {
+    pub fn add_and_store_user_mailbox<S: Storage>(
+        &mut self,
+        storage: &S,
+        user_id: ActorId,
+        message_id: MessageId,
+        mailbox_message: MailboxMessage,
+    ) {
         self.changed = true;
 
-        let r = self
-            .inner
-            .entry(user_id)
-            .or_default()
-            .insert(message_id, ValueWithExpiry { value, expiry });
-        debug_assert!(r.is_none())
+        let user_mailbox = if let Some(hash) = self.inner.get(&user_id) {
+            let mut user_mailbox = storage
+                .read_user_mailbox(*hash)
+                .expect("failed to read user mailbox from storage");
+
+            user_mailbox.add(message_id, mailbox_message);
+            user_mailbox
+        } else {
+            let mut user_mailbox = UserMailbox::default();
+            user_mailbox.add(message_id, mailbox_message.clone());
+            user_mailbox
+        };
+
+        let hash = storage.write_user_mailbox(user_mailbox);
+
+        let _ = self.inner.insert(user_id, hash);
     }
 
-    pub fn remove(
+    pub fn remove_and_store_user_mailbox<S: Storage>(
         &mut self,
+        storage: &S,
         user_id: ActorId,
         message_id: MessageId,
-    ) -> Option<ValueWithExpiry<u128>> {
-        let local_mailbox = self.inner.get_mut(&user_id)?;
-        let claimed_value = local_mailbox.remove(&message_id)?;
-
+    ) -> Option<MailboxMessage> {
         self.changed = true;
```

Comment:
now it causes bug. self.changed previously were only called if local/user mailbox for real contained value we're about to remove